File: 871.minimum-number-of-refueling-stops.cpp
/*
 * @lc app=leetcode id=871 lang=cpp
 *
 * [871] Minimum Number of Refueling Stops
 */
#include <bits/stdc++.h>

using std::vector;

// @lc code=start
class Solution {
public:
    // T: O(n * log(n))
    // S: O(n)
    int minRefuelStops(int target, int startFuel, vector<vector<int>>& stations) {
        if (startFuel >= target)  return 0;  // trivial
        
        const int n = stations.size();
        std::priority_queue<int> max_heap;
        int i = 0, reachable = startFuel, ans = 0;
        while (reachable < target) {
            while (i < n && stations[i][0] <= reachable) {
                // push all reachable fuels to the queue
                max_heap.push(stations[i++][1]);
            }
            if (max_heap.empty())  return -1;
            // Refuel with the maximum available one
            reachable += max_heap.top(); max_heap.pop();
            ++ans;
        }
        return ans;
    }
};
// @lc code=end

File: 871.minimum-number-of-refueling-stops.py
#
# @lc app=leetcode id=871 lang=python3
#
# [871] Minimum Number of Refueling Stops
#
import heapq
from typing import List

# @lc code=start
class Solution:
    # Greedy
    # T: O(n * log(n))
    # S: O(n)
    def minRefuelStops(self, target: int, startFuel: int, stations: List[List[int]]) -> int:
        if startFuel >= target: # trivial
            return 0

        n = len(stations)
        pq = []  # negate elements to simulate a max heap
        reachable = startFuel
        i = 0
        ans = 0
        while reachable < target:
            while i < n and stations[i][0] <= reachable:
                heapq.heappush(pq, -stations[i][1])
                i += 1
            
            if not len(pq):
                # Run out of stations
                return -1
            
            # Always refuel from the station with the greatest fuel.
            # Note that we can still refuel even if fuel tank is not empty.
            reachable += -heapq.heappop(pq)
            ans += 1
        
        return ans

# @lc code=end

